### ISSUE TITLE
Mobile HTML resources are not being updated when pushing new code to server

### ISSUE THREAD

**Issue by occasl** [meteor/meteor#5640]
I'm using Meteor 1.2.1 and I'm noticing that my client HTML resources on mobile devices are not being updated when I push new changes to the server. For instance, I have a file `app.html` under my `client` directory. Let's say I change something simple, like the text in a label inside the body, that change is not being push to my mobile clients, like it did when I used [Meteor Rider](https://github.com/zeroasterisk/MeteorRider).

The only way my clients will update is if I push a new IPA/APK. However, the HTML in question is not in there because the isomorphic nature of Meteor, so I'm confused why this won't update with a simple push of the new code to the server.

Can you please look into? This greatly impacts how we can deploy updates to our clients and defeats one of the main reasons we use Meteor which is to auto-update our clients without them having to install a new version from the App Store.

**Comment by martijnwalraven** [comment]
I'm not sure what is going on in your situation, but hot code push should definitely update your HTML templates.

I suspect there is a configuration issue. Are you sure your app is able to connect to the server? Do data updates work as expected?

**Comment by occasl** [comment]
Yes, everything else works as normal. It feels like the client is caching this file somehow. Although, I never had any problem with this before, could any of these plugins be causing that?  Maybe Appcache?

```
meteor-base             # Packages every Meteor app needs to have
mobile-experience       # Packages for a great mobile UX
mongo                   # The database Meteor supports right now
blaze-html-templates    # Compile .html files into Meteor Blaze views
session                 # Client-side reactive dictionary for your app
jquery                  # Helpful client-side library
tracker                 # Meteor's client-side reactive programming library

standard-minifiers      # JS/CSS minifiers run for production mode
es5-shim                # ECMAScript 5 compatibility for older browsers.
ecmascript              # Enable ECMAScript2015+ syntax in app code

underscore
accounts-base
http
skinnygeek1010:cordova-status@0.2.0
appcache
infinitedg:winston
mrt:underscore-string-latest

lepozepo:accounting
jquery-mobile
raix:push
jakobloekke:rabbitmq
momentjs:moment
```

Here is my mobile-config as well:
```
App.info({
    id: 'com.abc.appstore.mobile.approvals',
    name: 'Approval Hub',
    version: '1.1.2',
    description: 'Approval multiple IT system requests in one app'
});

// Set up resources such as icons and launch screens.
App.icons({
    'iphone': 'icons/icon-60.png'
//...
});

App.launchScreens({
    'iphone': 'splash/Default~iphone.png'
//...
});

// Add access rule for CORS access to meteor.local
App.accessRule("*");

// Set PhoneGap/Cordova preferences
App.setPreference('fullscreen', true);
App.setPreference('HideKeyboardFormAccessoryBar', true);
App.setPreference('Orientation', 'default');
App.setPreference('Orientation', 'all', 'ios');
```

**Comment by occasl** [comment]
I ruled out AppCache and a few other things. The lack of updates is very consistent no matter the changes I tried to make. I hooked my mobile device up to logcat and found some interesting errors in the logs:

```
11-10 12:35:14.401    9539-9539/? I/chromium﹕ [INFO:CONSOLE(12)] "METEOR CORDOVA DEBUG (meteor_cordova_loader.js) Error reading version file Error: Failed to resolve entry: file:///data/data/com.qualcomm.appstore.ea.mobile.approvals/files/meteor/version", source: file:///android_asset/www/meteor_cordova_loader.js (12)
11-10 12:35:14.401    9539-9539/? I/chromium﹕ [INFO:CONSOLE(12)] "METEOR CORDOVA DEBUG (meteor_cordova_loader.js) Couldn't load from the manifest, falling back to the bundled assets.", source: file:///android_asset/www/meteor_cordova_loader.js (12)
11-10 12:35:14.401    9539-9619/? W/meteor.cordova.updater﹕ startServer(/android_asset/www/application/,/android_asset/www/)
11-10 12:35:14.401    9539-9619/? W/meteor.cordova.updater﹕ Filesystem root not found; falling back to assets: /android_asset/www/application/
11-10 12:35:14.401    9539-9619/? D/meteor.cordova.updater﹕ Doing assetManager list on www
11-10 12:35:14.421    9539-9539/? I/chromium﹕ [INFO:CONSOLE(12)] "METEOR CORDOVA DEBUG (meteor_cordova_loader.js) Loading from url: http://meteor.local", source: file:///android_asset/www/meteor_cordova_loader.js (12)
11-10 12:35:14.421    9539-9580/? V/meteor.cordova.updater﹕ remapUri http://meteor.local/
11-10 12:35:14.421    9539-9580/? D/meteor.cordova.updater﹕ Doing assetManager list on www/application
11-10 12:35:14.431    9539-9580/? D/meteor.cordova.updater﹕ Found asset, but was directory: file:///android_asset/www/application/
11-10 12:35:14.431    9539-9580/? V/meteor.cordova.updater﹕ Assuming not a directory: index.html
11-10 12:35:14.431    9539-9580/? D/meteor.cordova.updater﹕ Remapping to file:///android_asset/www/application/index.html
11-10 12:35:14.441    9539-9580/? V/meteor.cordova.updater﹕ remapUri http://meteor.local/5705a73f81414e75615c1076b2690435270bd61b.css?meteor_css_resource=true
```

First it says it can't read version file, then can't load manifest so falls back to "bundled assets" and loads from http://meteor.local rather than my host server. I'm assuming bundled assets mean something loaded with the APK, which may explain the stale index.html but looking through the directory I don't see any old files. Any idea why it would be doing that?

**Comment by martijnwalraven** [comment]
Can you try deleting the app from the device and reinstalling?

**Comment by occasl** [comment]
I've tried that several times and I've tried deleting the cache on device as well but to no avail. The only way it will take any new changes if I do a `meteor build --server [my-meteor-server]` resulting in a new APK that I load on the device. It doesn't make any sense.

**Comment by martijnwalraven** [comment]
For some more context, Cordova apps on Meteor always load from `http://meteor.local`, even when the assets have been hot code pushed. It is basically equivalent to the `file://` URLs regular Cordova apps use to load from the file system, but intercepted by our plugin.

Not being able to read the version file is also not abnormal, this just means there are no hot code pushed versions yet. But hot code push should definitely kick in when you make changes.

Does hot code push work when you open the app in a regular browser?

**Comment by occasl** [comment]
Ok, thanks for the clarification. I do see that hot code push is working fine for Chrome on my macbook.

Further down in logcat, I also see this:

```
11-10 13:55:05.141  19935-19935/? I/chromium﹕ [INFO:CONSOLE(66)] "Uncaught Error: cannot call methods on tabs prior to initialization; attempted to call method 'option'", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (66)
11-10 13:55:05.481  19935-19935/? I/chromium﹕ [INFO:CONSOLE(51)] "WebSocket connection to 'wss://ma-meteor-demo.abc.com/sockjs/519/x1jbu0e_/websocket' failed: Error during WebSocket handshake: net::ERR_CONNECTION_RESET", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (51)
11-10 13:55:06.391  19935-19935/? I/chromium﹕ [INFO:CONSOLE(0)] "XMLHttpRequest cannot load https://ma-meteor-demo.abc.com/__cordova/manifest.json. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://meteor.local' is therefore not allowed access.", source: http://meteor.local/ (0)
11-10 13:55:06.391  19935-19935/? I/chromium﹕ [INFO:CONSOLE(152)] "METEOR CORDOVA DEBUG (autoupdate_cordova.js) Failed to download the manifest network undefined", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (152)
11-10 13:55:06.651  19935-19935/? I/chromium﹕ [INFO:CONSOLE(202)] "Subscribing to requests for: lsacco", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (202)
11-10 13:55:07.131  19935-19935/? I/chromium﹕ [INFO:CONSOLE(202)] "Mobile Approvals version loaded: 1.1.2.1", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (202)
11-10 13:55:10.111  19935-19935/? I/chromium﹕ [INFO:CONSOLE(202)] "****** Reconnected ******", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (202)
```

The WebSockets error I'm expecting because our infrastructure doesn't currently support it so I just expect it to fallback to XHR long poll, but the "No 'Access-Control-Allow-Origin' header" CORS error is a bit concerning. Nonetheless, it does seems to connect to the server for data. If I add records to Mongo the immediately show up in the mobile app.

**Comment by martijnwalraven** [comment]
The `No 'Access-Control-Allow-Origin' header` error is definitely the problem. The hot code push is triggered, but it fails to download the manifest that describes the assets to download. 

Not sure what is going on here. Is `https://ma-meteor-demo.abc.com` what you specified for the `--server` option? Did you specify a `Content-Security-Policy` meta tag by any chance? Also, does this happen both on iOS and Android?

**Comment by occasl** [comment]
Yes, I specify my server as part of the build. I don't set Content-Security-Policy and it happens on both iOS/Android. In my `mobile-config.js` I have the this:

```
// Add access rule for CORS access to meteor.local
App.accessRule("*");
```

And I also have the browser-policy plugin loaded with this config in my `server/server.js`:
```
    // Configure BrowserPolicy for CORS to work
    // https://github.com/meteor/meteor/issues/4132
    BrowserPolicy.content.allowSameOriginForAll();
    BrowserPolicy.content.allowOriginForAll('http://meteor.local');
    BrowserPolicy.content.allowOriginForAll('https://meteor.local');
    BrowserPolicy.content.allowOriginForAll('http://ma-meteor-demo.qualcomm.com');
    BrowserPolicy.content.allowOriginForAll('https://ma-meteor-demo.qualcomm.com');
    BrowserPolicy.content.allowOriginForAll('http://*.qualcomm.com');
    BrowserPolicy.content.allowOriginForAll('https://*.qualcomm.com');
    BrowserPolicy.content.allowEval();
    BrowserPolicy.framing.disallow();
```

The issue I reference in my comments seems to be outstanding for folks in spite of these recommendations.

**Comment by occasl** [comment]
I was able to get around the CORS error by adding the `webapp` package and adding this to me server.js (per CORS docs):

```
    WebApp.connectHandlers.use(function(req, res, next) {
        res.setHeader("Access-Control-Allow-Origin", "*");
        return next();
    });
```

Though after adding that, I now get this in logcat:
```
11-10 17:11:50.091    6893-6893/? I/chromium﹕ [INFO:CONSOLE(51)] "WebSocket connection to 'wss://ma-meteor-demo.abc.com/sockjs/050/eokhde8n/websocket' failed: Error during WebSocket handshake: net::ERR_CONNECTION_RESET", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (51)
11-10 17:11:51.591    6893-6893/? I/chromium﹕ [INFO:CONSOLE(152)] "METEOR CORDOVA DEBUG (autoupdate_cordova.js) Failed to download the manifest null <!DOCTYPE html>
    <html manifest="/app.manifest">
    <head>
    <link rel="stylesheet" type="text/css" class="__meteor-css__" href="/5705a73f81414e75615c1076b2690435270bd61b.css?meteor_css_resource=true">
    <script type="text/javascript">__meteor_runtime_config__ = JSON.parse(decodeURIComponent("%7B%22meteorRelease%22%3A%22METEOR%401.2.1%22%2C%22PUBLIC_SETTINGS%22%3A%7B%7D%2C%22ROOT_URL%22%3A%22https%3A%2F%2Fma-meteor-demo.abc.com%22%2C%22ROOT_URL_PATH_PREFIX%22%3A%22%22%2C%22autoupdateVersion%22%3A%226b7bf0ecb2ffc181b9639d0fbd10d56dc2e28f48%22%2C%22autoupdateVersionRefreshable%22%3A%222be9a4345110d0099f58ac84fa8dd00f0c615b1d%22%2C%22autoupdateVersionCordova%22%3A%22none%22%7D"));</script>
    <script type="text/javascript" src="/4d15640561b802f0eeb3a2e949d368f6b67b40d6.js?meteor_js_resource=true"></script>
    <title>Mobile Approvals</title>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,height=device-height,initial-scale=1, maximum-scale=1.0, user-scalable=no">
    <meta name="format-detection" content="telephone=no">
    </head>
    <body>
    </body>
    </html>
    ", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (152)
11-10 17:11:52.421    6893-6893/? I/chromium﹕ [INFO:CONSOLE(202)] "Subscribing to requests for: lsacco", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (202)
11-10 17:11:53.101    6893-6893/? I/chromium﹕ [INFO:CONSOLE(202)] "Mobile Approvals version loaded: 1.1.2.1", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (202)
11-10 17:11:54.221    6893-6893/? I/chromium﹕ [INFO:CONSOLE(202)] "****** Reconnected ******", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (202)
```

So help me understand what this manifest.json is and how it's produced because when I go to the filesystem of the server I do not see it anywhere. I think I would be able to serve the file now but for some reason it is not there.

**Comment by martijnwalraven** [comment]
I don't know why modifying `WebApp.connectHandlers` would cause this, but the server now seems to respond with `index.html` when requesting `manifest.json`. The manifest is not a stored file but gets generated in memory (see [webapp_server.js](https://github.com/meteor/meteor/blob/devel/packages/webapp/webapp_server.js)). Also, from the code [here](https://github.com/meteor/meteor/blob/devel/packages/webapp/webapp_server.js#L401-L403) it seems `Access-Control-Allow-Origin` is actually already being set specifically for `manifest.json`.

`BrowserPolicy` should set the `Content-Security-Policy` meta tag, among other things, but I'm not sure how this interacts with Cordova. Can you try removing the `browser-policy` package?

**Comment by martijnwalraven** [comment]
There definitely seems to be something weird going on. I wasn't aware https://github.com/meteor/meteor/issues/4132 was an active issue, so I need to investigate further. Thanks for bringing this to my attention!

**Comment by martijnwalraven** [comment]
Also, what version of Meteor are you running?

**Comment by occasl** [comment]
Removing `browser-policy` package yielded the same results. If I remove the handler to add the `Access-Control-Allow-Origin` header that put me back to not being able to connect again resulting in this error (reference 4 comments above):

```
11-11 12:03:36.381  25518-25518/? I/chromium﹕ [INFO:CONSOLE(152)] "METEOR CORDOVA DEBUG (autoupdate_cordova.js) Failed to download the manifest network undefined", source: http://meteor.local/c763e79132d7ceef70a369408b9aa12cc921394c.js?meteor_js_resource=true (152)
```

I'm running the latest version of Meteor 1.2.1.

**Comment by occasl** [comment]
Any news on this one?  We have another deployment and as far as I can tell I will need to push a new APK/IPA for everyone to get updated?  I don't understand why the client would still maintain the old code if I'm updating the server with new (HTML/JS) code. Where is the client caching it??

It also appears that I do have to set a CORS Header on our Proxy to allow for meteor.local:

```
Header set Access-Control-Allow-Origin "http:/meteor.local"
```

Maybe that has something to do with it?

**Comment by martijnwalraven** [comment]
This should be fixed in Meteor 1.3! A new beta is now available, please give it a try: https://github.com/meteor/meteor/issues/6266.

**Comment by georgexo** [comment]
@occasl I see its been 2 years but Im running Meteor 1.5 with cordova, seems to have the same issues you were having. I'm curious if a Meteor update is how to resolved the issue, OR did you find another way?

All help appreciated, thank you!
